definitions/ContentDefinitions.php

```diff
@@ -52,10 +52,12 @@ public static function getContentMetadata(Content $content)
             'updated_by' => $content->updatedBy ? UserDefinitions::getUserShort($content->updatedBy) : null,
             'updated_at' => $content->updated_at,
             'scheduled_at' => $content->scheduled_at,
-            'url' => $content->getUrl(true),
+            'url' => $content->contentcontainer_id ? $content->getUrl(true) : null,
```

Comment:
A content without a ContentContainer should also be able to have a URL.

If there is a problem here, we should rather fix it in the `getUrl()` method.

Reply:
> A content without a ContentContainer should also be able to have a URL.
> 
> If there is a problem here, we should rather fix it in the `getUrl()` method.

Maybe something like this would be a better solution?
```php
protected static function getContentMetadata($content)
{
    if ($content === null) {
        return null;
    }

    $metadata = [
        'id' => $content->id,
        'guid' => $content->guid,
        'object_model' => $content->object_model,
        'object_id' => $content->object_id,
        'visibility' => (int) $content->visibility,
        'state' => (int) $content->state,
        'archived' => (bool) $content->archived,
        'hidden' => (bool) $content->hidden,
        'pinned' => (bool) $content->pinned,
        'locked_comments' => (bool) $content->locked_comments,
        'created_by' => $content->createdBy ? UserDefinitions::getUserShort($content->createdBy) : null,
        'created_at' => $content->created_at,
        'updated_by' => $content->updatedBy ? UserDefinitions::getUserShort($content->updatedBy) : null,
        'updated_at' => $content->updated_at,
        'scheduled_at' => $content->scheduled_at,
        'contentcontainer_id' => $content->contentcontainer_id,
        'stream_channel' => $content->stream_channel
    ];

    // Get the URL if available
    if ($content->contentcontainer_id !== null) {
        $metadata['url'] = $content->getUrl(true);
    } else {
        $metadata['url'] = null;
    }

    return $metadata;
}
```

Another approach would be updating the core Content model class `getUrl()` to something like this;
```php
/**
 * Returns the URL of this content.
 *
 * By default, it returns the URL of the wall entry.
 *
 * Optionally, it's possible to create a custom `getUrl` method in the underlying
 * content model (e.g., Post) to override this behavior.
 * e.g., in case there is no wall entry available for this content.
 *
 * @param bool $scheme Whether to include the scheme (http/https) in the URL. Default is false.
 * @return string The URL of the content.
 * @throws \yii\db\Exception
 * @since 1.15
 */
public function getUrl($scheme = false)
{
    try {
        $polymorphicRelation = $this->getPolymorphicRelation();
        if ($polymorphicRelation && method_exists($polymorphicRelation, 'getUrl')) {
            return $polymorphicRelation->getUrl($scheme);
        }
    } catch (\Throwable $e) {
        Yii::error($e->getMessage(), 'content');
    }

    // Provide a default URL in case no specific URL is available
    if ($this->contentcontainer_id !== null) {
        // If content is associated with a ContentContainer, use PermaLink widget
        return \humhub\modules\content\widgets\PermaLink::widget(['content' => $this]);
    } else {
        // If content is not associated with a ContentContainer, generate a generic URL
        return \yii\helpers\Url::toRoute(['/content/perma', 'id' => $this->id], $scheme);
    }
}

```

Reply:
@ArchBlood What is wrong, to handle content without content container via `/content/perma`? There should be a conditions for such cases: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/controllers/PermaController.php#L53

Reply:
> @ArchBlood What is wrong, to handle content without content container via `/content/perma`? There should be a conditions for such cases: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/controllers/PermaController.php#L53

From looking at this, `$content->container` is used for an object that must not be `null` then later uses `createUrl()` to generate the URLs with `$urlParams` which equals a null array (i.e. `['contentId' => $id]`) when dealing with the Custom Pages module;

In the Custom Pages module the following is done where `getTargetModel()` is returning a `null` value therefore not allowing for `getContentUrl()` and throwing the error.

Within the PermaLink widget this isn't a possibility that is why it was used within `getUrl()` for the Content model class then later in the else we return the preferred method `Url::toRoute(['/content/perma', 'id' => $this->id], $scheme)` which covers everything else;

So in short `getTargetModel()` is returned as a nulled object which can't be done as `getContentUrl()` will throw the error when used in the ContentDefinitions class (i.e. `$content->getUrl()`)
```php
    /**
     * Returns the view url of this page.
     */
    public function getUrl()
    { 
        return $this->getTargetModel()->getContentUrl($this);
    }
`

Reply:
@ArchBlood Sorry, I don't understand the problem.

In principle, global content (without ContentContainer) can also have a perma URL via `Content::getUrl()`.

`Content::getUrl()` must not throw an error here,

If an error occurs here, we should fix the method and not skip the url in the REST module.